src/string/rabin-karp.md

```diff
@@ -7,16 +7,27 @@ e_maxx_link: rabin_karp
 # Rabin-Karp Algorithm for string matching
 
 This algorithm is based on the concept of hashing, so if you are not familiar with string hashing, refer to the [string hashing](string-hashing.md) article.
- 
 This algorithm was authored by Rabin and Karp in 1987.
 
-Problem: Given two strings - a pattern $s$ and a text $t$, determine if the pattern appears in the text and if it does, enumerate all its occurrences in $O(|s| + |t|)$ time.
+Problem: Given two strings - a length $m$ pattern to find in a length $n$ text, find all matches in $\Theta(m+n) = \Theta(n)$ average time.
 
-Algorithm: Calculate the hash for the pattern $s$.
-Calculate hash values for all the prefixes of the text $t$.
-Now, we can compare a substring of length $|s|$ with $s$ in constant time using the calculated hashes.
-So, compare each substring of length $|s|$ with the pattern. This will take a total of $O(|t|)$ time.
-Hence the final complexity of the algorithm is $O(|t| + |s|)$: $O(|s|)$ is required for calculating the hash of the pattern and $O(|t|)$ for comparing each substring of length $|s|$ with the pattern.
+The naive solution is to simply check all length $m$ substrings in the $n$ length text, but that would take $\Theta(mn)$ time.
+
+The algorithm uses the concept of a "rolling hash", in which we compute a hash for the pattern and another for the first substring in $\Theta(m)$ time, but then use a special hash function that lets us go from one to the next in constant time. We compare the hash to our known hash of the pattern, and if they are the same then we likely have a match.
+
+## The hash function
+
+The trick is the special polynomial hash function, also known as a Rabin fingerprint. It is defined as 
+
+$$h(S[i..j]) = S[i] x^{m-1} + S[i+1] x^{m-2} + \cdots + S[j-1] x + S[j]$$
+
+This is a polynomial in variable $x$, and $x$ is chosen to be as large as the alphabet, so that we have a unique hash for every fixed-length substring. For example, for a string of bytes, we can take $x = 256$. However, the resulting hash value is large, so we work modulo a large prime $p$. This creates a small chance of hash collisions that needs to be handled. In the worst case of a hash collision every time, the performance degrades to $\Theta(nm)$. 
```

Comment:
> In the worst case of a hash collision every time, the performance degrades to $\Theta(nm)$.

It doesn't, hash collisions only lead to false matches.

Besides, we should add:

1. What should be randomized (either $p$ or $x$ should do);
2. What is the expected probability of a false match due to hash collision.

But I'm generally not sure whether it should be covered right here, or in the general hashing article (in which case this paragraph should link to a corresponding paragraph).

Reply:
The way it is presented usually is that for every hash match we still check for correctness. And the time complexity is in CLRS.

Reply:
Then it diverges from the provided implementation, and searching for all matches of $a^{n/2}$ in $a^n$ would be quadratic due to the checks for correctness...

Reply:
Yes, the assumption in CLRS is that the number of valid matches is low. I'm not opposed to rewriting it to assume a hash collision never happens, but I would make that explicit.

Reply:
I'd say what we need is not the assumption that a collision never happens, but rather a proper estimate on the probability of it happening, so that people can adjust their randomization range and e.g. number of mods to be checked to ensure a probability that they're fine with.